cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -384,20 +384,69 @@ func generateInitramfsMounts() error {
 }
 
 func unlockIfEncrypted(name string) (string, error) {
-	// TODO:UC20: will need to unseal key to unlock LUKS here
 	device := filepath.Join("/dev/disk/by-label", name)
-	keyfile := filepath.Join(boot.InitramfsUbuntuBootDir, name+".keyfile.unsealed")
-	if osutil.FileExists(keyfile) {
+	encdev := device + "-enc"
+	if osutil.FileExists(encdev) {
 		// TODO:UC20: snap-bootstrap should validate that <name>-enc is what
 		//            we expect (and not e.g. an external disk), and also that
 		//            <name> is from <name>-enc and not an unencrypted partition
 		//            with the same name (LP #1863886)
-		cmd := exec.Command("/usr/lib/systemd/systemd-cryptsetup", "attach", name, device+"-enc", keyfile)
-		cmd.Env = os.Environ()
-		cmd.Env = append(cmd.Env, "SYSTEMD_LOG_TARGET=console")
-		if output, err := cmd.CombinedOutput(); err != nil {
-			return "", osutil.OutputErr(output, err)
+		sealedKeyPath := filepath.Join(boot.InitramfsEncryptionKeyDir, name+".keyfile.sealed")
+		if err := unlockEncryptedPartition(name, encdev, sealedKeyPath, "", ""); err != nil {
+			return device, fmt.Errorf("cannot unlock %s: %v", name, err)
 		}
```

Comment:
we should do that whenever there's a TPM right (even if secure boot is not enabled), right?

this one is quite important and needs to be covered here or an immediate follow up

it also probably means that the CheckKeySealingEnable should probably actually return flags about whether TPM and/or secure boot are enabled/present

Reply:
Yes, all of the interactions with the TPM during boot with the exception of unsealing should happen whenever there is a TPM device detected, regardless of whether secure boot is enabled or there is an encrypted volume to unlock.

Reply:
Obviously, if secure boot is disabled then you can't unseal any keys anyway because we won't create policies that allow unsealing when secure boot is disabled. So, technically you could omit interactions with the TPM. But that will create extra code paths and make it more likely to introduce bugs, so it's simpler just to interact with the TPM unconditionally.

Reply:
Will cover this on an immediate follow up.